README.md

# asiakastieto

Nodejs **Asiakastieto integration** for our projects.

This library is developed for our internal purposes, but you may find it useful if you work with Asiakastieto in similar environment too.

## Installation

  **Requirements:** Suomen Asiakastieto agreement and credentials, NodeJS.

  ```bash
  npm install asiakastieto --save
  ```

  or using Yarn:

  ```bash
  yarn add asiakastieto
  ```

## Usage

```javascript

const Asiakastieto = require('asiakastieto');

const asiakastietoConfig = {
  userid: process.env.USERID,
  passwd: process.env.PASSWD,
  shaKey: process.env.SHA_KEY,
  requestPromise: require('request-promise')
};

// do customer default check
const asiakastieto = new Asiakastieto(config);
const params = {
  enduser: 'ccccc',
  idnumber: process.env.ID_NUMBER,
  lang: 'EN',
  sequence: 25
};
const url = asiakastieto.buildDefaultCheckUrl(params);
const data = await asiakastieto.doRequestAndParseXML(url);

```


## More documentation

* [Asiakastieto](https://www.asiakastieto.fi/)
* [ILOapps.es](https://iloapps.es/)


## Support

THe package is provided 'AS IS'. Usage of this package is with your own care and responsibility. ILO APPLICATIONS SL does not provide any warranty or support for this package. The package is intended for our own projects and we use it in production/live environment. If we see any problems in our projects, we will update and fix accordingly.

For any bug report and improvement ideas, we are happy to receive them at support (at sign) iloapps.es.. :)



CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [0.3.0](https://github.com/iloapplications/asiakastieto/compare/0.2.0...0.3.0)

> 12 February 2022

- Bump y18n from 3.2.1 to 3.2.2 [`#6`](https://github.com/iloapplications/asiakastieto/pull/6)
- Bump yargs-parser from 5.0.0 to 5.0.1 [`#7`](https://github.com/iloapplications/asiakastieto/pull/7)
- Bump hosted-git-info from 2.7.1 to 2.8.9 [`#4`](https://github.com/iloapplications/asiakastieto/pull/4)
- Bump copy-props from 2.0.4 to 2.0.5 [`#5`](https://github.com/iloapplications/asiakastieto/pull/5)
- Bump validator from 11.0.0 to 13.7.0 [`#3`](https://github.com/iloapplications/asiakastieto/pull/3)
- Bump mixin-deep from 1.3.1 to 1.3.2 [`#2`](https://github.com/iloapplications/asiakastieto/pull/2)
- Bump handlebars from 4.1.2 to 4.7.7 [`#1`](https://github.com/iloapplications/asiakastieto/pull/1)
- Update package information and dvv basic info [`1f9b6c2`](https://github.com/iloapplications/asiakastieto/commit/1f9b6c2fe32cddaac39f79fe2e282fe6c944c81b)
- npm i [`986c9c2`](https://github.com/iloapplications/asiakastieto/commit/986c9c22fbd752e7e416c654299dcf9e124a25c2)
- Update readme [`aff3342`](https://github.com/iloapplications/asiakastieto/commit/aff33426299935f7a8c1b9be2e2b11919d8b0ca4)

#### 0.2.0

> 12 February 2022

- First version ready, using Asiakastieto consumer5/REST interface [`6d35280`](https://github.com/iloapplications/asiakastieto/commit/6d352803e9ae2d8b3856ad4d9205f2ba3ece6f7c)
- Add initial tests [`296e152`](https://github.com/iloapplications/asiakastieto/commit/296e152f2961a3135eadc04c51ac867cf58fc725)
- Configuration files [`6c0598a`](https://github.com/iloapplications/asiakastieto/commit/6c0598a97742f01817c5929f2ffd59344f67b309)


test/index.js
'use strict';

const should = require('should');
const chai = require('chai');
const validator = require('validator');
const chaiAsPromised = require('chai-as-promised');
const expect = chai.expect;
const sinon = require('sinon');
const sinonChai = require('sinon-chai');
const MockDate = require('mockdate');
const fs = require('fs');
chai.use(sinonChai);
chai.use(chaiAsPromised);
const Asiakastieto = require('../index');
const urlOptions = {
  protocols: ['https'],
  require_valid_protocol: true,
  require_protocol: true,
  require_tld: false // for localhost
};

const apiUrl = 'https://demo.asiakastieto.fi/services/consumer5/REST';
const stringRange = (start, end, zeros) => Array.from(Array(end - start + 1), (_, i) => {
  return (start + i).toString().padStart(zeros, '0');
});
const checksumDemo = '71BCA0FDF4926E784A1C86B64E528435305B5FFD0952C6B1834F234D44CAC69674E0D7A95A484F1708D82DC623F24D3F331732619E5ED7E3237DA139A8D00643';
const authenticationFailureXml = fs.readFileSync('./test/xml/authentication_failure.xml');
const versionFailureXml = fs.readFileSync('./test/xml/version_not_supported.xml');
const genericErrorXml = fs.readFileSync('./test/xml/generic_error.xml');
const defaultCheckResponseXml = fs.readFileSync('./test/xml/default_check_response.xml');

let config, basicParams;
beforeEach(() => {
  MockDate.reset();
  config = {
    userid: 'testusername',
    passwd: 'pwd123',
    shaKey: 'ABXEYYUEOVJCMEUEWUJDJ',
    isProd: false,
    requestPromise: function () {}
  };
  basicParams = {
    enduser: 'cccccc',
    idnumber: '23423423423',
    sequence: (Math.floor(Math.random() * (99999 - 1 + 1)) + 1),
    lang: 'FI',
    ccenter: '123456789'
  };
});

/*  YYYYMMDDHHMMSSXXTZNNNNNN where
      YYYY = year in four digits
      MM = month 01 - 12
      DD = day of the month 01 - 31
      HH = hour 00 - 23
      MM = minutes 00 - 59
      SS = seconds 00 - 59
      XX = hundredths of a second 00 - 99
      TZN = time zone correction in relation to GMT.
      In Finland“ % 2 B02”(plus has to be included as UTF - 8 hexadecimal value % 2 B).The offset from
      GMT is always presented in relation to local
      normal time, daylight saving time(DST) is not
      added to the correction value.
      NNNNN = consecutive number
    */
function validateTimestamp (timestamp, sequence) {
  expect(stringRange(2010, 2100)).to.include(timestamp.substr(0, 4));
  expect(stringRange(1, 12, 2)).to.include(timestamp.substr(4, 2));
  expect(stringRange(1, 31, 2)).to.include(timestamp.substr(6, 2));
  expect(stringRange(0, 23, 2)).to.include(timestamp.substr(8, 2));
  expect(stringRange(0, 59, 2)).to.include(timestamp.substr(10, 2));
  expect(stringRange(0, 59, 2)).to.include(timestamp.substr(12, 2));
  expect(stringRange(0, 999, 3)).to.include(timestamp.substr(14, 3));

  expect(timestamp.substr(-5)).to.equal(sequence.toString().padStart(5, '0'));
}

function validateBasicUrlParams (urlObject, config, params) {
  expect(urlObject.protocol + '//' + urlObject.host + urlObject.pathname).to.equal(apiUrl);
  expect(urlObject.searchParams.get('passwd')).to.equal(config.passwd);
  expect(urlObject.searchParams.get('userid')).to.equal(config.userid);
  expect(urlObject.searchParams.get('version')).to.equal('2018');
  expect(urlObject.searchParams.get('format')).to.equal('xml');
  expect(urlObject.searchParams.get('enduser')).to.equal(params.enduser);
  expect(urlObject.searchParams.get('idnumber')).to.equal(params.idnumber);
  expect(urlObject.searchParams.get('lang')).to.equal(params.lang);
  validateTimestamp(urlObject.searchParams.get('timestamp'), params.sequence);
  expect(urlObject.searchParams.get('checksum').length).to.equal(checksumDemo.length);
  expect(urlObject.searchParams.get('ccenter')).to.equal(params.ccenter);
}

describe('Asiakastieto', function () {
  describe('Constructor', () => {
    it('Should return an error if requestPromise not configured', () => {
      delete config.requestPromise;
      expect(() => new Asiakastieto(config)).to.throw(Error, /requestPromise/);
    });
    it('Should return an error when userid is missing', () => {
      delete config.userid;
      expect(() => new Asiakastieto(config)).to.throw(Error, /userid/);
    });
    it('Should return an error when userid is not 12 chars', () => {
      config.userid = 'testuserid';
      expect(() => new Asiakastieto(config)).to.throw(Error, /length must be 12 chars/);
    });
    it('Should return an error when passwd is missing', () => {
      delete config.passwd;
      expect(() => new Asiakastieto(config)).to.throw(Error, /passwd/);
    });
    it('Should return an error when passwd length is not valid', () => {
      config.passwd = 'password1234';
      expect(() => new Asiakastieto(config)).to.throw(Error, /passwd is too large/);
    });
    it('Should return an error when shaKey is missing', () => {
      delete config.shaKey;
      expect(() => new Asiakastieto(config)).to.throw(Error, /shaKey/);
    });
    it('Should return an error when isProd is not a valid boolean', () => {
      config.isProd = 'true';
      expect(() => new Asiakastieto(config)).to.throw(Error, /isProd/);
    });
  });

  describe('Build timestamp', () => {
    it('Should return a formatted timestamp that matches required format(+2 GMT): YYYYMMDDHHMMSSXXTZNNNNNN', () => {
      MockDate.set(new Date(), +120);
      // i.e: 2008021514330000%2B0200001
      const consecutiveNumber = (Math.floor(Math.random() * (99999 - 1 + 1)) + 1);
      const timestamp = new Asiakastieto(config).buildTimestamp(consecutiveNumber);
      console.log('timestamp: ', timestamp);

      validateTimestamp(timestamp, consecutiveNumber);
      expect(timestamp.substr(17, 5)).to.equal('%2B02');
    });

    it('Should return a formatted timestamp that matches required format(+0 GMT): YYYYMMDDHHMMSSXXTZNNNNNN', () => {
      // i.e: 2008021514330000%2B0200001
      const consecutiveNumber = (Math.floor(Math.random() * (99999 - 1 + 1)) + 1);
      const timestamp = new Asiakastieto(config).buildTimestamp(consecutiveNumber);
      console.log('timestamp: ', timestamp);

      validateTimestamp(timestamp, consecutiveNumber);
      expect(timestamp.substr(17, 2)).to.equal('00');
    });
  });

  describe('Calculate checksum', () => {
    /*
      The checksum (the SHA-512 hash value) is presented as a hexadecimal character string of 128
      characters, in which letter symbols are capital letters.
      userid&enduser&timestamp&key&
    */
    it('Should calculate a valid SHA-512 checksum', () => {
      const params = {
        enduser: 'cccccc',
        timestamp: '2005110112264400%2B0200001'
      };
      const checksum = new Asiakastieto(config).calculateChecksum(params);
      expect(checksum.length).to.equal(checksumDemo.length);
    });
  });

  describe('Basic parameters', () => {
    it('Should throw an error if idnumber is missing ', () => {
      delete basicParams.idnumber;
      expect(() => new Asiakastieto(config).createUrlWithBasicParameters(basicParams)).to.throw(Error, /idnumber is mandatory/);
    });
    it('Should throw an error if idnumber format is not valid', () => {
      basicParams.idnumber = '123456789';
      expect(() => new Asiakastieto(config).createUrlWithBasicParameters(basicParams)).to.throw(Error, /idnumber format is not valid/);
    });
    it('Should throw an error if enduser is missing', () => {
      delete basicParams.enduser;
      expect(() => new Asiakastieto(config).createUrlWithBasicParameters(basicParams)).to.throw(Error, /enduser is mandatory/);
    });
    it('Should throw an error if enduser format is not valid', () => {
      basicParams.enduser = 'enduserinvalidformat';
      expect(() => new Asiakastieto(config).createUrlWithBasicParameters(basicParams)).to.throw(Error, /enduser format is not valid/);
    });
    it('Should throw an error if sequence is missing', () => {
      delete basicParams.sequence;
      expect(() => new Asiakastieto(config).createUrlWithBasicParameters(basicParams)).to.throw(Error, /sequence/);
    });
    it('Should throw an error if sequence is not a valid number', () => {
      basicParams.sequence = 'a1324';
      expect(() => new Asiakastieto(config).createUrlWithBasicParameters(basicParams)).to.throw(Error, /Must be a valid number/);
    });
    it('Should throw an error if language is not valid (FI, EN, SV)', () => {
      basicParams.lang = 'ES';
      expect(() => new Asiakastieto(config).createUrlWithBasicParameters(basicParams)).to.throw(Error, /locale is not valid/);
    });
    it('Should throw an error if ccenter is too long', () => {
      basicParams.ccenter = 'costcenterinvalidformat';
      expect(() => new Asiakastieto(config).createUrlWithBasicParameters(basicParams)).to.throw(Error, /ccenter is not valid/);
    });

    it('Should build a valid URL with basic parameters', () => {
      const urlObject = new Asiakastieto(config).createUrlWithBasicParameters(basicParams);
      // Check specific values per service are not set
      expect(urlObject.searchParams.get('reqmsg')).to.equal(null);
      expect(urlObject.searchParams.get('request')).to.equal(null);
      expect(urlObject.searchParams.get('qtype')).to.equal(null);
      expect(urlObject.searchParams.get('level')).to.equal(null);
      expect(urlObject.searchParams.get('purpose')).to.equal(null);
    });
  });

  describe('Consumer information', () => {
    /* Example URL:
      https://demo.asiakastieto.fi/services/consumer5/REST?version=2018
        &lang=FI
        &userid=nnnnnnnnnnnn
        &passwd=xxxxxx
        &enduser=ccccc
        &reqmsg=CONSUMER
        &idnumber=ddmmyyAnnnX
        &purpose=1
        &qtype=41
        &format=xml
        &timestamp=2004070713434000%2B0200001
        &checksum=xxxxxxxxxxxxxxxxxxxxxxxxxxx&
    */
    it('Should build a valid URL for payment default check', () => {
      const url = new Asiakastieto(config).buildDefaultCheckUrl(basicParams);
      const urlObject = new URL(url);
      validateBasicUrlParams(urlObject, config, basicParams);

      // Check specific values per service
      expect(urlObject.searchParams.get('reqmsg')).to.equal('CONSUMER');
      expect(urlObject.searchParams.get('request')).to.equal('H');
      expect(urlObject.searchParams.get('qtype')).to.equal('41');
      expect(urlObject.searchParams.get('level')).to.equal('');
      expect(urlObject.searchParams.get('purpose')).to.equal('1');
    });

    /* Example URL:
      https://demo.asiakastieto.fi/services/consumer5/REST?version=2018
        &lang=FI
        &userid=nnnnnnnnnnnn
        &passwd=xxxxxx
        &enduser=ccccc
        &reqmsg=CONSUMER
        &idnumber=ddmmyyAnnnX
        &purpose=1
        &qtype=41
        &level=1
        &format=xml
        &timestamp=2004070713434000%2B0200001
        &checksum=xxxxxxxxxxxxxxxxxxxxxxxxxxx&
     */
    it('Should build a valid URL for Credit information check', () => {
      const url = new Asiakastieto(config).buildInformationCheckUrl(basicParams);
      const urlObject = new URL(url);
      validateBasicUrlParams(urlObject, config, basicParams);

      // Check specific values per service
      expect(urlObject.searchParams.get('reqmsg')).to.equal('CONSUMER');
      expect(urlObject.searchParams.get('request')).to.equal('H');
      expect(urlObject.searchParams.get('qtype')).to.equal('42');
      expect(urlObject.searchParams.get('level')).to.equal('');
      expect(urlObject.searchParams.get('purpose')).to.equal('1');
    });
  });

  describe('Decision support service', () => {
    /* Example URL:
      https://demo.asiakastieto.fi/services/consumer5/REST?version=2018
        &userid=nnnnnnnnnnnn
        &passwd= xxxxx
        &enduser=ccccc
        &idnumber=ddmmyyAnnnX
        &reqmsg=DECISION
        &qtype=02
        &model=MALLI1
        &C6_input.code%5B0%5D=A05
        &C6_input.value%5B0%5D=1000
        &C6_input.code%5B1%5D=A08
        &C6_input.value%5B1%5D=123456
        &request=H
        &format=xml
        &timestamp=2005110112264400%2B0 200001
        &checksum=xxxxxxxx&
    */

    it('Should build a valid URL with mandatory parameters', () => {
      const url = new Asiakastieto(config).buildDecisionSupportServiceUrl(basicParams);
      const urlObject = new URL(url);
      validateBasicUrlParams(urlObject, config, basicParams);
      expect(urlObject.searchParams.get('reqmsg')).to.equal('DECISION');
      expect(urlObject.searchParams.get('request')).to.equal('H');
      expect(urlObject.searchParams.get('qtype')).to.equal('02');
    });

    it('Should build a valid URL with full parameters (model, A05, A08)', () => {
      basicParams.model = 'MALL';
      basicParams.creditAmount = 1200;
      basicParams.customerKey = '1234';
      const url = new Asiakastieto(config).buildDecisionSupportServiceUrl(basicParams);
      const urlObject = new URL(url);
      validateBasicUrlParams(urlObject, config, basicParams);
      const leftBracketUtf8 = '%5B';
      const rightBracketUtf8 = '%5D';
      // Check specific values per service
      expect(urlObject.searchParams.get('reqmsg')).to.equal('DECISION');
      expect(urlObject.searchParams.get('request')).to.equal('H');
      expect(urlObject.searchParams.get('qtype')).to.equal('02');
      expect(urlObject.searchParams.get('model')).to.equal(basicParams.model);
      expect(urlObject.searchParams.get('C6_input.code%5B0%5D')).to.equal('A05');
      expect(urlObject.searchParams.get('C6_input.value%5B0%5D', basicParams.creditAmount.toString()));
      expect(urlObject.searchParams.get('C6_input.code%5B0%5D', 'A08'));
      expect(urlObject.searchParams.get('C6_input.value%5B0%5D', basicParams.customerKey.toString()));
    });
  });

  describe('Simppeli', () => {
    /* Example URL:
      https://demo.asiakastieto.fi/services/consumer5/REST?userid=nnnnnnnnnnnn
        &passwd=XXXXXX
        &enduser=cccccc
        &idnumber=ddmmyyAnnnX
        &purpose=1
        &request=H
        &lang=FImodel=SIMP1
        &version=2018
        &qtype=04
        &reqmsg=DECISION
        &format=xml
        &timestamp=2008021514330000%2B0200001
        &checksum=xxxxxxxxxxx&
    */

    it('Should build a valid URL without model', () => {
      const url = new Asiakastieto(config).buildSimppeliUrl(basicParams);
      const urlObject = new URL(url);
      validateBasicUrlParams(urlObject, config, basicParams);

      // Check specific values per service
      expect(urlObject.searchParams.get('reqmsg')).to.equal('DECISION');
      expect(urlObject.searchParams.get('request')).to.equal('H');
      expect(urlObject.searchParams.get('qtype')).to.equal('04');
    });

    it('Should build a valid URL with model included', () => {
      basicParams.model = 'SIMP1';
      const url = new Asiakastieto(config).buildSimppeliUrl(basicParams);
      const urlObject = new URL(url);
      validateBasicUrlParams(urlObject, config, basicParams);

      expect(urlObject.searchParams.get('model')).to.equal(basicParams.model);
    });
  });

  describe('Nordic Person information', () => {
    it('Should throw an error if country is not valid (SWE, NOR, DNK)', () => {
      basicParams.country = 'FIN';
      expect(() => new Asiakastieto(config).buildNordicPersonInfoUrl(basicParams)).to.throw(Error, /country is not valid/);
    });

    it('Should build a valid URL', () => {
      basicParams.country = 'DNK';
      const url = new Asiakastieto(config).buildNordicPersonInfoUrl(basicParams);
      const urlObject = new URL(url);
      validateBasicUrlParams(urlObject, config, basicParams);

      // Check specific values per service
      expect(urlObject.searchParams.get('reqmsg')).to.equal('CONSUMER');
      expect(urlObject.searchParams.get('qtype')).to.equal('12');
      expect(urlObject.searchParams.get('country')).to.equal(basicParams.country);
      expect(urlObject.searchParams.get('qmsg')).to.equal('P3');
    });
  });

  describe('Do GET Request', () => {
    it('Should thrown an error if errorMessage(errorCode, errorText) is found in XML response', () => {
      sinon.stub(config, 'requestPromise').resolves(authenticationFailureXml);
      expect(new Asiakastieto(config).doRequestAndParseXML(apiUrl)).to.be.rejectedWith(Error);
    });

    it('Should thrown an error if XML error found', () => {
      sinon.stub(config, 'requestPromise').resolves(versionFailureXml);
      expect(new Asiakastieto(config).doRequestAndParseXML(apiUrl)).to.be.rejectedWith(Error);
    });

    it('Should get and parse Generic error XML response to JSON', () => {
      sinon.stub(config, 'requestPromise').resolves(genericErrorXml);
      expect(new Asiakastieto(config).doRequestAndParseXML(apiUrl)).to.be.rejectedWith(Error);
    });

    it('Should get and parse consumer default check response from XML to JSON', async () => {
      sinon.stub(config, 'requestPromise').resolves(defaultCheckResponseXml);
      const data = await new Asiakastieto(config).doRequestAndParseXML(apiUrl);
      expect(validator.isJSON(JSON.stringify(data))).to.equal(true);
    });
  });
});


gulpfile.js
'use strict';
// npm install gulp gulp-mocha gulp-util gulp-nodemon

const gulp = require('gulp');
const mocha = require('gulp-mocha');
const gutil = require('gulp-util');
const nodemon = require('gulp-nodemon');
const stubby = require('gulp-stubby-server');
const jshint = require('gulp-jshint');
const stylish = require('jshint-stylish');
const eslint = require('gulp-eslint');
const istanbul = require('gulp-istanbul');
const lintspaces = require('gulp-lintspaces');
const babel = require('gulp-babel');

const fs = require('fs');

gulp.task('pre-mocha', function() {
  return gulp.src(['./index.js'])
    .pipe(babel({ presets: ['@babel/env'] }))
    .pipe(istanbul({ includeUntested: true }))
    .pipe(istanbul.hookRequire());
});

gulp.task('tabpreventer', function() {
  return gulp.src(['index.js'])
    .pipe(lintspaces({
      indentation: 'spaces',
      spaces: 2,
      ignores: [
        'js-comments',
        'c-comments'
      ]
    }))
    .pipe(lintspaces.reporter());
});

gulp.task('mocha-watch', function() {
  return gulp.src(['test/**/*.js'])
    .pipe(mocha())
    .on('error', gutil.log);
});

gulp.task('mocha', gulp.series('pre-mocha', function() {
  return gulp.src(['test/**/*.js'])
    .pipe(mocha())
    .on('error', gutil.log);
}));

gulp.task('watch', function() {
  gulp.watch(['*.js','test/**'], ['mocha-watch']);
});

gulp.task('hint', function () {
  return gulp.src(['index.js'])
    .pipe(jshint())
    .pipe(jshint.reporter(stylish))
    .on('error', function (error) {
       console.error(String(error));
  });
});

gulp.task('watch-lint', function() {
  gulp.watch(['index.js'], ['lint']);
});

gulp.task('lint', function() {
  return gulp.src(['index.js']).pipe(eslint())
  .pipe(eslint.format())
  // Brick on failure to be super strict
  .pipe(eslint.failOnError());
});

gulp.task('pre-push', gulp.series('tabpreventer', 'lint', 'hint', 'mocha'));

gulp.task('default', gulp.series(function (done) {
  // task code here
  done();
}));


examples/simppeli.js
const Asiakastieto = require('../index'); // require('asiakastieto')
const assert = require('assert');

assert.ok(process.env.USERID, 'Please specify a USERID env var');
assert.ok(process.env.PASSWD, 'Please specify a PASSWD env var');
assert.ok(process.env.SHA_KEY, 'Please specify a SHA_KEY env var');

const config = {
  userid: process.env.USERID,
  passwd: process.env.PASSWD,
  shaKey: process.env.SHA_KEY,
  requestPromise: require('request-promise')
};

run().catch(error => {
  console.error(error.stack);
  process.exit(0);
});

async function run () {
  const asiakastieto = new Asiakastieto(config);
  const params = {
    enduser: 'ccccc',
    idnumber: '020770-0952', // Only for testing (MHL4)
    lang: 'EN',
    sequence: 27,
    model: 'SIMP1'
  };
  const url = asiakastieto.buildSimppeliUrl(params);
  console.log(url);
  const data = await asiakastieto.doRequestAndParseXML(url);

  console.log(JSON.stringify(data, null, 2));

  process.exit(0);
}


examples/dvv.basic.data.js
const Asiakastieto = require('../index'); // require('asiakastieto')
const assert = require('assert');

assert.ok(process.env.USERID, 'Please specify a USERID env var');
assert.ok(process.env.PASSWD, 'Please specify a PASSWD env var');
assert.ok(process.env.SHA_KEY, 'Please specify a SHA_KEY env var');

const config = {
  userid: process.env.USERID,
  passwd: process.env.PASSWD,
  shaKey: process.env.SHA_KEY,
  requestPromise: require('request-promise')
};

run().catch(error => {
  console.error(error.stack);
  process.exit(0);
});

async function run () {
  const asiakastieto = new Asiakastieto(config);
  const params = {
    enduser: 'testtesttest',
    idnumber: '010100-123D',
    lang: 'EN',
    sequence: 27
  };
  const url = asiakastieto.buildDvvBasicDataUrl(params);
  console.log(url);
  const data = await asiakastieto.doRequestAndParseXML(url);

  console.log(JSON.stringify(data, null, 2));

  process.exit(0);
}


index.js
'use strict';

const crypto = require('crypto');
const parser = require('xml2js');
const util = require('util');

const stagUrl = 'https://demo.asiakastieto.fi/services/consumer5/REST';
const prodUrl = 'https://www.asiakastieto.fi/services/consumer5/REST';

function validateConfig (config) {
  if (!config.requestPromise) {
    throw new Error('Configuration: requestPromise is mandatory');
  }
  if (config.isProd && typeof (config.isProd) !== 'boolean') {
    throw new Error('Configuration: isProd must be a boolean');
  }
  if (!config.userid) {
    throw new Error('Configuration: userid is mandatory');
  }
  if (config.userid.length !== 12) {
    throw new Error('Configuration: userid is invalid (length must be 12 chars)');
  }
  if (!config.passwd) {
    throw new Error('Configuration: passwd is mandatory');
  }
  if (config.passwd.length > 10) {
    throw new Error('Configuration: passwd is too large (maximum 10 chars)');
  }
  if (!config.shaKey) {
    throw new Error('Configuration: shaKey is missing');
  }
}

function validateMandatory (params) {
  if (!params.idnumber) {
    throw new Error('Validation: idnumber is mandatory');
  }
  if (params.idnumber.length > 11) {
    throw new Error('Validation: idnumber format is not valid (too long)');
  }
  if (params.idnumber.length < 11) {
    throw new Error('Validation: idnumber format is not valid (too short)');
  }
  if (!params.enduser) {
    throw new Error('Validation: enduser is mandatory');
  }
  if (params.enduser.length > 12) {
    throw new Error('Validation: enduser format is not valid (too long)');
  }
  if (!params.sequence || isNaN(params.sequence)) {
    throw new Error('Validation: sequence is not valid. Must be a valid number');
  }
}

function validateOptional (params) {
  if (params.lang && !['FI', 'EN', 'SV'].includes(params.lang)) {
    throw new Error('Validation: locale is not valid (only FI, EN or SV are valid)');
  }
  if (params.ccenter && params.ccenter.length > 20) {
    throw new Error('Validation: ccenter is not valid (too long)');
  }
}

function validateDecisionSupportService (params) {
  if (params.model && params.model.length > 6) {
    throw new Error('Validation: model format is not valid (too long)');
  }
  if (params.creditAmount && (isNaN(params.creditAmount) || params.creditAmount < 0)) {
    throw new Error('Validation: creditAmount must be a valid positive number');
  }
  if (params.customerKey && params.customerKey.length > 30) {
    throw new Error('Validation: customerKey format is not valid (too large)');
  }
}

function validateNordicPersonInformation (params) {
  if (params.country && !['SWE', 'NOR', 'DNK'].includes(params.country)) {
    throw new Error('Validation: country is not valid (only SWE, NOR or DNK are allowed');
  }
}

module.exports = function Asiakastieto (config) {
  const apiUrl = config.isProd ? prodUrl : stagUrl;
  validateConfig(config);

  function buildTimestamp (consecutiveNumber) {
    const date = new Date();
    let offset = date.getTimezoneOffset();

    if (offset !== 0) {
      offset = (offset < 0 ? '%2D' : '%2B') + (Math.abs(offset) / 60).toString().padStart(2, '0');
    } else {
      offset = offset.toString().padStart(2, '0');
    }

    return date.getFullYear().toString() +
      (date.getMonth() + 1).toString().padStart(2, '0') +
      date.getDate().toString().padStart(2, '0') +
      date.getHours().toString().padStart(2, '0') +
      date.getMinutes().toString().padStart(2, '0') +
      date.getSeconds().toString().padStart(2, '0') +
      date.getMilliseconds().toString().padStart(3, '0') +
      offset +
      consecutiveNumber.toString().padStart(5, '0');
  }

  function calculateChecksum (userid, enduser, timestamp, secret) {
    const dataString = userid + '&' + enduser + '&' + timestamp + '&' + secret + '&';

    return (crypto.createHash('sha512')).update(dataString, 'utf8').digest('hex').toUpperCase();
  }

  function createUrlWithBasicParameters (params) {
    validateMandatory(params);
    validateOptional(params);

    const url = new URL(apiUrl);
    // Add mandatory parameters
    url.searchParams.append('userid', config.userid);
    url.searchParams.append('passwd', config.passwd);
    url.searchParams.append('enduser', params.enduser);
    url.searchParams.append('idnumber', params.idnumber);
    url.searchParams.append('version', 2018);
    url.searchParams.append('format', 'xml');
    const timestamp = buildTimestamp(params.sequence);
    url.searchParams.append('timestamp', timestamp);
    url.searchParams.append('checksum', calculateChecksum(config.userid, params.enduser, timestamp, config.shaKey));
    // Add voluntary parameters
    if (params.lang) {
      url.searchParams.append('lang', params.lang);
    }
    if (params.ccenter) {
      url.searchParams.append('ccenter', params.ccenter);
    }

    return url;
  }

  function buildDefaultCheckUrl (params) {
    const url = createUrlWithBasicParameters(params);

    url.searchParams.append('reqmsg', 'CONSUMER');
    url.searchParams.append('request', 'H');
    url.searchParams.append('qtype', 41);
    url.searchParams.append('level', '');
    url.searchParams.append('purpose', 1);

    return url.href;
  }

  function buildInformationCheckUrl (params) {
    const url = createUrlWithBasicParameters(params);

    url.searchParams.append('reqmsg', 'CONSUMER');
    url.searchParams.append('request', 'H');
    url.searchParams.append('qtype', 42);
    url.searchParams.append('level', '');
    url.searchParams.append('purpose', 1);

    return url.href;
  }

  function buildDecisionSupportServiceUrl (params) {
    const url = createUrlWithBasicParameters(params);
    validateDecisionSupportService(params);
    url.searchParams.append('reqmsg', 'DECISION');
    url.searchParams.append('request', 'H');
    url.searchParams.append('qtype', '02');

    if (params.model) {
      url.searchParams.append('model', params.model);
    }

    const leftBracketStr = '%5B';
    const rightBracketStr = '%5D';
    if (params.creditAmount) {
      url.searchParams.append(`C6_input.code${leftBracketStr}0${rightBracketStr}`, 'A05');
      url.searchParams.append(`C6_input.value${leftBracketStr}0${rightBracketStr}`, params.creditAmount);
    }
    if (params.customerKey) {
      url.searchParams.append(`C6_input.code${leftBracketStr}1${rightBracketStr}`, 'A08');
      url.searchParams.append(`C6_input.value${leftBracketStr}1${rightBracketStr}`, params.customerKey);
    }

    return url.href;
  }

  function buildSimppeliUrl (params) {
    const url = createUrlWithBasicParameters(params);

    url.searchParams.append('reqmsg', 'DECISION');
    url.searchParams.append('request', 'H');
    url.searchParams.append('qtype', '04');
    if (params.model) {
      url.searchParams.append('model', params.model);
    }

    return url.href;
  }

  function buildNordicPersonInfoUrl (params) {
    const url = createUrlWithBasicParameters(params);
    validateNordicPersonInformation(params);
    url.searchParams.append('reqmsg', 'CONSUMER');
    url.searchParams.append('qtype', '12');
    url.searchParams.append('purpose', 1);
    url.searchParams.append('country', params.country);
    url.searchParams.append('qmsg', 'P3');

    return url.href;
  }

  function buildDvvBasicDataUrl (params) {
    const url = createUrlWithBasicParameters(params);

    url.searchParams.append('reqmsg', 'CONSUMER');
    url.searchParams.append('request', 'H');
    url.searchParams.append('qtype', 34);
    // url.searchParams.append('level', '');
    url.searchParams.append('purpose', 1);

    return url.href;
  }

  async function doRequestAndParseXML (url) {
    const rpOptions = { method: 'GET', uri: url };
    const response = await config.requestPromise(rpOptions);

    const parserOptions = {
      tagNameProcessors: [name => name.replace(/ns[2,3,4]:/gi, '')],
      explicitArray: false
    };

    const data = await util.promisify(parser.parseString.bind(parser))(response, parserOptions);
    const consumerResponse = data && data.response && data.response.consumerResponse;
    if (!consumerResponse) {
      throw new Error('Unable to parse XML response');
    }
    if (consumerResponse.errorMessage) {
      throw new Error('Error in XML response: ' + JSON.stringify(consumerResponse.errorMessage, null, 2));
    }

    return consumerResponse;
  }

  return {
    calculateChecksum,
    buildTimestamp,
    createUrlWithBasicParameters,
    buildDefaultCheckUrl,
    buildInformationCheckUrl,
    buildDecisionSupportServiceUrl,
    buildSimppeliUrl,
    buildNordicPersonInfoUrl,
    buildDvvBasicDataUrl,
    doRequestAndParseXML
  };
};
